### ISSUE TITLE
Race condition when creating realtime collection

### ISSUE THREAD

**Issue by yahor-filipchyk** [scullxbones/akka-persistence-mongo#213]
When multiple queries start executing on different nodes at the same time, some start failing with an error
```
Command failed with error 48 (NamespaceExists): 'a collection 
'my_db.akka_persistence_realtime already exists' on server mongodb-server.com:27017. 
The full response is 
{ 
  "ok" : 0.0, 
  "errmsg" : "a collection 'my_db.akka_persistence_realtime' already exists",
  "code" : 48, 
  "codeName" : "NamespaceExists" 
}
```
I think this becomes especially a problem because the reference to the realtime collection is stored in plugin as `lazy val`, wrapped in a `Future`. So even though we run a stream in an actor and the actor restarts when the stream fails, the stream will keep failing because it keeps getting the same old error from the same `Future` instance. If I understand the code correctly.
I guess it would be very difficult to coordinate collection creation across nodes so allowing the collection to recover from failure sounds like a reasonable solution.

**Comment by yahor-filipchyk** [comment]
Hi @scullxbones,
I don't think it is related to #199. The realtime collection was declared like this forever:
```scala
  private[mongodb] lazy val realtime: C = {
    cappedCollection(realtimeCollectionName)(concurrent.ExecutionContext.global)
  }
```
Perhaps the issue has something to do with implementation of `cappedCollection` in the official driver extension? In any case if something goes wrong in `cappedCollection` call, there is no way to recover, is there?
I guess the behavior will be different with Casbah because 
```scala
  // Collection type
  type C = MongoCollection
```
so error in initializing realtime collection will probably blow the whole thing up, which might be better

**Comment by yufei-cai** [comment]
I ran into this error for suffixed event journal collections as well. The event journal collections are not capped. Specifically, I see
```
com.mongodb.MongoCommandException: Command failed with error 48 (NamespaceExists): 'a collection 'journal_name@suffix' already exists'
```
when 10 actors in a 3-node Akka cluster persist events with identical journal suffix simultaneously into a single-master MongoDB hosted on AWS.

**Comment by scullxbones** [comment]
@yufei-cai - by that are you referring to this part, or something else?  Meaning the losers of the race that is set up by a multi-node deploy aren't quietly losing.

> Additionally it looks like the new scala official driver isn't handling the `namespace already exists` error.

**Comment by yufei-cai** [comment]
I mean when the persistent actors calls "persist" simultaneously, `EventSourced.onPersistFailure` gets called with the "NamespaceExists" error as cause on losers of the race.

> Additionally it looks like the new scala official driver isn't handling the `namespace already exists` error.

Someone will have to handle the error. The persistence plugin isn't the worst place to do it for a consistent behavior across driver choices.

**Comment by scullxbones** [comment]
Ok yep, that's what I was confirming.  I agree the `official scala` flavor of plugin should handle the race loss quietly like the `reactive mongo` flavor.

There's two issues already in this ticket, so was looking to make sure that your case was one of the two, thanks!

**Comment by yufei-cai** [comment]
@scullxbones How far are you into a solution? I got permissions to work on a pull request.

**Comment by scullxbones** [comment]
@yufei-cai - I haven't touched this yet.  I look forward to a PR, thanks!

**Comment by scullxbones** [comment]
@yufei-cai let me know what you're thinking as far as a schedule, I've got #222 sitting in master and am wondering whether I should wait on the PR or just go ahead and cut a release.  Thanks!

**Comment by yufei-cai** [comment]
The PR will be ready tomorrow at the latest. Please feel free to go ahead with the release.

**Comment by scullxbones** [comment]
Ok so missed the fact that the fix doesn't address the original (but smaller) issue with `realtime`.  Opening that into another issue.

**Comment by scullxbones** [comment]
Suffixed collection fixed in `2.2.3`
